LES client panics after syncing is done

#### System information

Geth version: `Version: 1.10.18-unstable
Git Commit: 25bd17d725db5378b3611ac077506dc09e940a77`
OS & Version: OSX

#### What happened

Immediately after the light client finished syncing from a local light server it crashed.

```sh
DEBUG[03-30|09:21:27.166] Fetching batch of headers                id=0eaf38fb12f42c09 conn=dyndial count=192  fromnum=6,623,214 skip=0   reverse=false
TRACE[03-30|09:21:27.167] Light Ethereum message arrived           id=0eaf38fb12f42c09 conn=dyndial code=3  bytes=16
TRACE[03-30|09:21:27.167] Received block header response message   id=0eaf38fb12f42c09 conn=dyndial
INFO [03-30|09:21:27.167] Imported new block headers               count=2    elapsed=1.633ms   number=6,623,213 hash=9cd49a..f77956 age=13h22m31s
DEBUG[03-30|09:21:27.168] Synchronise finished                     elapsed=2m22.876s
INFO [03-30|09:21:27.558] Storing bloom trie                       section=201 head=d764fc53973d2d80887b70196ab2bbd1ee802878a347bcacd3ea881424ccb4fd root=041dfbb3335d21f8d9d2dd8d58b2e45f469eab888b2575879074bc46069e1690 compression=0.460
INFO [03-30|09:21:28.264] Storing CHT                              section=201 head=d764fc53973d2d80887b70196ab2bbd1ee802878a347bcacd3ea881424ccb4fd root=06a4b2487785d1914b65dae0e4f8284570798a19f74f717c0432f5822862c15c


panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x1c0 pc=0x486e884]
                                                                                                                                                                                                                          goroutine 205 [running]:
github.com/ethereum/go-ethereum/les.(*lightFetcher).mainloop(0xc0003aedc0)
        github.com/ethereum/go-ethereum/les/fetcher.go:445 +0x21a4
created by github.com/ethereum/go-ethereum/les.(*lightFetcher).start
        github.com/ethereum/go-ethereum/les/fetcher.go:198 +0x87
```

--- a/les/fetcher.go
+++ b/les/fetcher.go
@@ -446,8 +446,6 @@ func (f *lightFetcher) mainloop() {
 				// is not found. It can happen the original head is before the
 				// checkpoint while the synced headers are after it. In this
 				// case there is no ancestor between them.
-				// TODO(rjl493456442) it's quite expensive for searching the
-				// ancestor in this case, try to avoid it in the first place.
 				if ancestor == nil {
 					ancestor = f.chain.Genesis().Header()
 				}

--- a/les/fetcher.go
+++ b/les/fetcher.go
@@ -441,6 +441,16 @@ func (f *lightFetcher) mainloop() {
 			if ulc {
 				head := f.chain.CurrentHeader()
 				ancestor := rawdb.FindCommonAncestor(f.chaindb, origin, head)
+
+				// Recap the ancestor with genesis header in case the ancestor
+				// is not found. It can happen the original head is before the
+				// checkpoint while the synced headers are after it. In this
+				// case there is no ancestor between them.
+				// TODO(rjl493456442) it's quite expensive for searching the
+				// ancestor in this case, try to avoid it in the first place.
+				if ancestor == nil {
+					ancestor = f.chain.Genesis().Header()
+				}
 				var untrusted []common.Hash
 				for head.Number.Cmp(ancestor.Number) > 0 {
 					hash, number := head.Hash(), head.Number.Uint64()
@@ -449,6 +459,9 @@ func (f *lightFetcher) mainloop() {
 					}
 					untrusted = append(untrusted, hash)
 					head = f.chain.GetHeader(head.ParentHash, number-1)
+					if head == nil {
+						break // all the synced headers will be dropped
+					}
 				}
 				if len(untrusted) > 0 {
 					for i, j := 0, len(untrusted)-1; i < j; i, j = i+1, j-1 {
@@ -514,7 +527,7 @@ func (f *lightFetcher) requestHeaderByHash(peerid enode.ID) func(common.Hash) er
 	}
 }
 
-// requestResync invokes synchronisation callback to start syncing.
+// startSync invokes synchronisation callback to start syncing.
 func (f *lightFetcher) startSync(id enode.ID) {
 	defer func(header *types.Header) {
 		f.syncDone <- header

--- a/les/fetcher_test.go
+++ b/les/fetcher_test.go
@@ -25,6 +25,7 @@ import (
 	"github.com/ethereum/go-ethereum/core"
 	"github.com/ethereum/go-ethereum/core/rawdb"
 	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/light"
 	"github.com/ethereum/go-ethereum/p2p/enode"
 	"github.com/ethereum/go-ethereum/params"
 )
@@ -152,6 +153,7 @@ func TestTrustedAnnouncementsLes2(t *testing.T) { testTrustedAnnouncement(t, 2) 
 func TestTrustedAnnouncementsLes3(t *testing.T) { testTrustedAnnouncement(t, 3) }
 
 func testTrustedAnnouncement(t *testing.T, protocol int) {
+	//log.Root().SetHandler(log.LvlFilterHandler(log.LvlDebug, log.StreamHandler(os.Stderr, log.TerminalFormat(true))))
 	var (
 		servers   []*testServer
 		teardowns []func()
@@ -159,16 +161,28 @@ func testTrustedAnnouncement(t *testing.T, protocol int) {
 		ids       []string
 		cpeers    []*clientPeer
 		speers    []*serverPeer
+
+		config       = light.TestServerIndexerConfig
+		waitIndexers = func(cIndexer, bIndexer, btIndexer *core.ChainIndexer) {
+			for {
+				cs, _, _ := cIndexer.Sections()
+				bts, _, _ := btIndexer.Sections()
+				if cs >= 2 && bts >= 2 {
+					break
+				}
+				time.Sleep(10 * time.Millisecond)
+			}
+		}
 	)
-	for i := 0; i < 10; i++ {
-		s, n, teardown := newTestServerPeer(t, 10, protocol)
+	for i := 0; i < 4; i++ {
+		s, n, teardown := newTestServerPeer(t, int(2*config.ChtSize+config.ChtConfirms), protocol, waitIndexers)
 
 		servers = append(servers, s)
 		nodes = append(nodes, n)
 		teardowns = append(teardowns, teardown)
 
 		// A half of them are trusted servers.
-		if i < 5 {
+		if i < 2 {
 			ids = append(ids, n.String())
 		}
 	}
@@ -185,6 +199,18 @@ func testTrustedAnnouncement(t *testing.T, protocol int) {
 			teardowns[i]()
 		}
 	}()
+
+	// Register the assembled checkpoint as hardcoded one.
+	head := servers[0].chtIndexer.SectionHead(0)
+	cp := &params.TrustedCheckpoint{
+		SectionIndex: 0,
+		SectionHead:  head,
+		CHTRoot:      light.GetChtRoot(servers[0].db, 0, head),
+		BloomRoot:    light.GetBloomTrieRoot(servers[0].db, 0, head),
+	}
+	c.handler.checkpoint = cp
+	c.handler.backend.blockchain.AddTrustedCheckpoint(cp)
+
 	// Connect all server instances.
 	for i := 0; i < len(servers); i++ {
 		sp, cp, err := connect(servers[i].handler, nodes[i].ID(), c.handler, protocol, true)
@@ -218,9 +244,9 @@ func testTrustedAnnouncement(t *testing.T, protocol int) {
 		}
 		verifyChainHeight(t, c.handler.fetcher, expected)
 	}
-	check([]uint64{1}, 1, func() { <-newHead })   // Sequential announcements
-	check([]uint64{4}, 4, func() { <-newHead })   // ULC-style light syncing, rollback untrusted headers
-	check([]uint64{10}, 10, func() { <-newHead }) // Sync the whole chain.
+	check([]uint64{1}, 1, func() { <-newHead })                                                                       // Sequential announcements
+	check([]uint64{config.ChtSize + config.ChtConfirms}, config.ChtSize+config.ChtConfirms, func() { <-newHead })     // ULC-style light syncing, rollback untrusted headers
+	check([]uint64{2*config.ChtSize + config.ChtConfirms}, 2*config.ChtSize+config.ChtConfirms, func() { <-newHead }) // Sync the whole chain.
 }
 
 func TestInvalidAnnouncesLES2(t *testing.T) { testInvalidAnnounces(t, lpv2) }

--- a/les/ulc_test.go
+++ b/les/ulc_test.go
@@ -55,7 +55,7 @@ func testULCAnnounceThreshold(t *testing.T, protocol int) {
 			ids       []string
 		)
 		for i := 0; i < len(testcase.height); i++ {
-			s, n, teardown := newTestServerPeer(t, 0, protocol)
+			s, n, teardown := newTestServerPeer(t, 0, protocol, nil)
 
 			servers = append(servers, s)
 			nodes = append(nodes, n)
@@ -132,10 +132,11 @@ func connect(server *serverHandler, serverId enode.ID, client *clientHandler, pr
 }
 
 // newTestServerPeer creates server peer.
-func newTestServerPeer(t *testing.T, blocks int, protocol int) (*testServer, *enode.Node, func()) {
+func newTestServerPeer(t *testing.T, blocks int, protocol int, indexFn indexerCallback) (*testServer, *enode.Node, func()) {
 	netconfig := testnetConfig{
 		blocks:    blocks,
 		protocol:  protocol,
+		indexFn:   indexFn,
 		nopruning: true,
 	}
 	s, _, teardown := newClientServerEnv(t, netconfig)

